WalletWasabi.Tests/UnitTests/WabiSabi/Integration/WabiSabiHttpApiIntegrationTests.cs

```diff
@@ -126,6 +130,78 @@ public async Task SoloCoinJoinTestAsync(long[] amounts, long[]? outputs = null)
 			await roundStateUpdater.StopAsync(CancellationToken.None);
 		}
 
+		[Fact]
+		public async Task MultiClientsCoinJoinTestAsync()
+		{
+			const int NumberOfParticipants = 20;
+			const int NumberOfCoinsPerParticipant = 2;
+			int expectedInputNumber = NumberOfParticipants * NumberOfCoinsPerParticipant;
+
+			var node = await TestNodeBuilder.CreateForHeavyConcurrencyAsync();
+			var rpc = node.RpcClient;
+
+			var httpClient = _apiApplicationFactory.WithWebHostBuilder(builder =>
+			{
+				builder.ConfigureServices(services =>
+				{
+					// Instruct the coodinator DI container to use these two scoped
+					// services to build everything (wabisabi controller, arena, etc)
+					services.AddScoped<IRPCClient>(s => rpc);
+					services.AddScoped<WabiSabiConfig>(s => new WabiSabiConfig
+					{
+						MaxRegistrableAmount = Money.Coins(500m),
+						MaxInputCountByRound = expectedInputNumber,
+						ConnectionConfirmationTimeout = TimeSpan.FromSeconds(20 * expectedInputNumber),
+						OutputRegistrationTimeout = TimeSpan.FromSeconds(20 * expectedInputNumber),
+					});
+				});
+			}).CreateClient();
+
+			// Create the API client
+			var apiClient = _apiApplicationFactory.CreateWabiSabiHttpApiClient(httpClient);
+
+			// At the end of the test a coinjoin transaction has to be created and broadcasted.
+			var transactionCompleted = new TaskCompletionSource<Transaction>();
+
+			// Total test timeout.
+			using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(20 * expectedInputNumber));
+			cts.Token.Register(() => transactionCompleted.TrySetCanceled(), useSynchronizationContext: false);
+
+			var participants = Enumerable
+				.Range(0, NumberOfParticipants)
+				.Select(_ => new Participant(rpc, apiClient))
+				.ToArray();
+
+			foreach (var participant in participants)
+			{
+				await participant.InitializeAsync(NumberOfCoinsPerParticipant, cts.Token);
+			}
+			using var dummyKey = new Key();
+			await rpc.GenerateToAddressAsync(101, dummyKey.PubKey.GetAddress(ScriptPubKeyType.Segwit, rpc.Network));
+
+			var tasks = participants.Select(x => x.StartParticipatingAsync(cts.Token)).ToArray();
+
+			while ((await rpc.GetRawMempoolAsync()).Length == 0)
+			{
+				if (cts.IsCancellationRequested)
+				{
+					throw new TimeoutException("CoinJoin was not propagated.");
+				}
+
+				await Task.Delay(500);
+
+				if (tasks.FirstOrDefault(t => t.IsFaulted)?.Exception is { } exc)
+				{
+					throw exc;
+				}
+			}
+			var mempool = await rpc.GetRawMempoolAsync();
+			var coinjoin = await rpc.GetRawTransactionAsync(mempool.Single());
+
+			Assert.True(coinjoin.Outputs.Count >= expectedInputNumber);
+			Assert.True(coinjoin.Inputs.Count == expectedInputNumber);
+		}
```

Comment:
Can you add `await node.TryStopAsync(false);` with a try/f ?

Reply:
it worked! thank you.